src/Control/Effect/Sum.hs

```diff
@@ -20,26 +22,59 @@ instance (HFunctor f, HFunctor g) => HFunctor (f :+: g)
 instance (Effect f, Effect g)     => Effect   (f :+: g)
 
 
-class Member (sub :: (* -> *) -> (* -> *)) sup where
-  inj :: sub m a -> sup m a
-  prj :: sup m a -> Maybe (sub m a)
+type Member (sub :: (* -> *) -> (* -> *)) sup = MemberAt (PathTo sub sup) sub sup
 
-instance Member sub sub where
-  inj = id
-  prj = Just
+inj :: forall sub m a sup . Member sub sup => sub m a -> sup m a
+inj = inj' @(PathTo sub sup)
 
-instance {-# OVERLAPPABLE #-} Member sub (sub :+: sup) where
-  inj = L . inj
-  prj (L f) = Just f
-  prj _     = Nothing
-
-instance {-# OVERLAPPABLE #-} Member sub sup => Member sub (sub' :+: sup) where
-  inj = R . inj
-  prj (R g) = prj g
-  prj _     = Nothing
+prj :: forall sub m a sup . Member sub sup => sup m a -> Maybe (sub m a)
+prj = prj' @(PathTo sub sup)
 
 
 -- | Construct a request for an effect to be interpreted by some handler later on.
 send :: (Member effect sig, Algebra sig m) => effect m a -> m a
 send = alg . inj
 {-# INLINE send #-}
+
+
+type family FromJust (maybe :: Maybe a) :: a where
+  FromJust ('Just a) = a
+
+type family (left :: Maybe k) <|> (right :: Maybe k) :: Maybe k where
+  'Just a <|> _       = 'Just a
+  _       <|> 'Just b = 'Just b
+  _       <|> _       = 'Nothing
+
+type family Prepend (s :: j -> k) (ss :: Maybe j) :: Maybe k where
+  Prepend s ('Just ss) = 'Just (s ss)
+  Prepend _ 'Nothing   = 'Nothing
+
+data L a
+data R a
```

Comment:
I ended up going with this instead of

```haskell
data Side = L | R
type Path = [Side]
```

because it makes it easier to restrict the kinds.